KAFKA-5150 reduce lz4 decompression overhead

xvrl opened the pull request:
- reuse decompression buffers in consumer Fetcher
- switch lz4 input stream to operate directly on ByteBuffers
- avoids performance impact of catching exceptions when reaching the end of legacy record batches
- more tests with both compressible / incompressible data, multiple
  blocks, and various other combinations to increase code coverage
- fixes bug that would cause exception instead of invalid block size
  for invalid incompressible blocks
- fixes bug if incompressible flag is set on end frame block size

Overall this improves LZ4 decompression performance by up to 40x for small batches.
Most improvements are seen for batches of size 1 with messages on the order of ~100B.
We see at least 2x improvements for for batch sizes of < 10 messages, containing messages < 10kB

This patch also yields 2-4x improvements on v1 small single message batches for other compression types.

Full benchmark results can be found here
https://gist.github.com/xvrl/05132e0643513df4adf842288be86efd

xvrl committed:
small batch decompression benchmark

xvrl committed:
KAFKA-5150 reduce lz4 decompression overhead

- reuse decompression buffers, keeping one per thread
- switch lz4 input stream to operate directly on ByteBuffers
- more tests with both compressible / incompressible data, multiple
  blocks, and various other combinations to increase code coverage
- fixes bug that would cause EOFException instead of invalid block size
  for invalid incompressible blocks

ijuma reviewed (commented):
Thanks for the PR, great results! Why not include the benchmarks in the jmh-benchmarks module?

xvrl commented:
@ijuma sure, I can add the benchmarks, wasn't aware of the module

ijuma commented on clients/src/main/java/org/apache/kafka/common/record/AbstractLegacyRecordBatch.java:
```diff
@@ -291,12 +290,12 @@ public AbstractLegacyRecordBatch nextBatch() throws IOException {
         private final byte wrapperMagic;
 
-        private DeepRecordsIterator(AbstractLegacyRecordBatch wrapperEntry, boolean ensureMatchingMagic, int maxMessageSize) {
+        protected DeepRecordsIterator(AbstractLegacyRecordBatch wrapperEntry, boolean ensureMatchingMagic, int maxMessageSize) {
```
Are the `private` -> `protected` changes still needed?

xvrl replied:
we don't need them anymore, indeed

ijuma commented on clients/src/main/java/org/apache/kafka/common/record/CompressionType.java:
```diff
@@ -96,9 +117,9 @@ public OutputStream wrapForOutput(ByteBufferOutputStream buffer, byte messageVer
 
         @Override
-        public InputStream wrapForInput(ByteBufferInputStream buffer, byte messageVersion) {
+        public InputStream wrapForInput(ByteBuffer buffer, byte messageVersion) {
```
For consistency, it may make sense for `wrapForOutput` to take a `ByteBuffer` too.

xvrl replied:
sure

xvrl replied:
seems like this is not as straightforward as it seems. MemoryRecordsBuilder relies on having access to the underlying ByteBuffer held by the ByteBufferOutputStream, which would break if the buffer is expanded in ByteBufferOutputStream

ijuma replied:
Good point, can we please add a comment explaining the reason for the inconsistency?

xvrl replied:
done

ijuma commented on clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java:
```diff
@@ -210,8 +209,7 @@ public int partitionLeaderEpoch() {
         buffer.position(RECORDS_OFFSET);
-        final DataInputStream stream = new DataInputStream(compressionType().wrapForInput(
-                new ByteBufferInputStream(buffer), magic()));
+        final DataInputStream stream = new DataInputStream(compressionType().wrapForInput(buffer, magic()));
```
We could change `DefaultRecord.readFrom` to take a `DataInput` (instead of `DataInputStream`), change `KafkaLZ4BlockInputStream` to implement `DataInput` and then only wrap if the returned value is not already a `DataInput`. That would remove a layer of indirection and if it's possible to implement `readFully` more efficiently in `KafkaLZ4BlockInputStream`, then it could be a win. If you have time, it may be worth a try.

xvrl replied:
I'd prefer to do that in a follow-up PR in trunk, since I'd also like to backport this to 0.10.2.x

xvrl replied:
I was also thinking, if we avoided readFully altogether, we could avoid the EOFException penalty when reaching the end of the batch on legacy records and get a further 3x improvement for small legacy batches.

ijuma replied:
Hmm, `DataInputStream.readFully` only throws an exception if we ask it to read past the end of the InputStream. So supposedly, if we fix the underlying InputStream, it's enough either way. The following PR does that:

https://github.com/apache/kafka/pull/2025/files#diff-eaa7e4414f285da2ff8e4508456078d2L192

ijuma commented on clients/src/main/java/org/apache/kafka/common/record/KafkaLZ4BlockInputStream.java:
```diff
@@ -16,46 +16,56 @@
+    private final ByteBuffer in;
+
+    private ByteBuffer theBuffer;
+    private final ByteBuffer decompressionBuffer;
```
It wasn't clear to me why we needed both `theBuffer` and `decompressionBuffer`. It seems that we rely on the fact that `theBuffer` is only populated after the first `readBlock` to implement `available` correctly. Is that the only difference? If so, would it not be simpler to have a single buffer that is allocated for the first time in `readBlock`?

xvrl replied:
`theBuffer` is also pointed directly to the input buffer when a block is not compressed to avoid copying bytes.

ijuma replied:
I had noticed that, but didn't think through how that could work if we had a single buffer variable. OK, seems simplest to have two buffers. Can we add a comment explaining this?

xvrl replied:
will do

ijuma commented on clients/src/main/java/org/apache/kafka/common/record/KafkaLZ4BlockInputStream.java:
```diff
@@ -260,5 +292,4 @@ public synchronized void reset() throws IOException {
     public boolean markSupported() {
```
`mark()` and `reset()` are synchronized and probably should not be.

xvrl replied:
hmm, not sure why they were in the first place

ijuma replied:
Copy and paste, I bet.

ijuma commented on jmh-benchmarks/src/main/java/org/apache/kafka/common/record/DeepRecordsIteratorBenchmark.java:
```diff
@@ -0,0 +1,124 @@
+    private CompressionType type = CompressionType.NONE;
+
+    @Param(value = {"1", "2"})
+    byte messageVersion = MAGIC_VALUE_V1;
```
I think the default should be `CURRENT_MAGIC_VALUE`.

xvrl replied:
fixed

ijuma commented on jmh-benchmarks/src/main/java/org/apache/kafka/common/record/DeepRecordsIteratorBenchmark.java:
```diff
@@ -0,0 +1,124 @@
+    }
+
+    @Param(value = {"LZ4", "SNAPPY"})
+    private CompressionType type = CompressionType.NONE;
```
Why did you remove `NONE` from this list? It seems useful to have the uncompressed baseline.

xvrl replied:
When using `MemoryRecords.readableRecords(...)` we don't technically test the Compression.NONE "compression" method, but test a different code path, so it's not really comparable if you care about an upper bound on how fast decompression could be.

xvrl replied:
I added it back so we could at least see what it means in practice though

ijuma replied:
It's not the same code path, but that's OK. The comparison I'm looking for is a bit like the one in the LZ4 GitHub page where it compares memcpy with lz4 decompression (https://github.com/lz4/lz4).

ijuma commented on jmh-benchmarks/src/main/java/org/apache/kafka/common/record/DeepRecordsIteratorBenchmark.java:
```diff
@@ -0,0 +1,124 @@
+    final Random random = new Random(0);
+    final int batchCount = 5000;
+
+    private int maxBatchSize = 10;
```
Seems like more fields could be final in this class.

xvrl replied:
fixed

ijuma commented on jmh-benchmarks/src/main/java/org/apache/kafka/common/record/DeepRecordsIteratorBenchmark.java:
```diff
@@ -0,0 +1,124 @@
+
+    private ByteBuffer createBatch(int batchSize) {
+        byte[] value = new byte[messageSize];
+        final ByteBuffer buf = ByteBuffer.allocate((messageSize + 8 + 4) * batchSize);
```
You should probably use `AbstractRecords.sizeInBytesUpperBound`.

xvrl replied:
good to know

ijuma commented on jmh-benchmarks/src/main/java/org/apache/kafka/common/record/DeepRecordsIteratorBenchmark.java:
```diff
@@ -0,0 +1,124 @@
+    @Warmup(iterations = 5)
+    @Benchmark
+    public void measureSingleMessage(Blackhole bh) throws IOException {
+        for (Record record : new ByteBufferLogInputStream(theBuffer.duplicate(), Integer.MAX_VALUE).nextBatch()) {
```
To make the benchmark less dependent on implementation details, it would be better if it used `MemoryRecords.readableRecords(...).batches()`. Same for the other benchmark.

xvrl replied:
changed.

ijuma commented on jmh-benchmarks/src/main/java/org/apache/kafka/common/record/DeepRecordsIteratorBenchmark.java:
```diff
@@ -0,0 +1,124 @@
+import static org.apache.kafka.common.record.RecordBatch.MAGIC_VALUE_V1;
+
+@State(Scope.Benchmark)
+public class DeepRecordsIteratorBenchmark {
```
Maybe this should be `RecordBatchIterationBenchmark`?

xvrl replied:
fixed

ijuma commented on clients/src/main/java/org/apache/kafka/common/record/CompressionType.java:
```diff
@@ -156,7 +177,8 @@ else if (LZ4.name.equals(name))
     private static class LZ4Constructors {
         static final MethodHandle INPUT = findConstructor(
                 "org.apache.kafka.common.record.KafkaLZ4BlockInputStream",
```
We followed the same approach for Snappy and LZ4 to deal with the possibility that the libraries are not in the classpath because it's not supported in a given platform. However, we are using Kafka classes for the block input stream and block output stream. As such, we can probably reference the constructor directly. We won't invoke the constructor unless LZ4 is configured and it's OK to fail in that case.

Does that make sense?

xvrl replied:
makes sense, as far as I know lz4 falls back to a pure-java version anyway, so it should be safe regardless (unless there are other platform issues that are unrelated to native code)

xvrl replied:
note however that we store the LZ4SafeDecompressor as a static field, so loading the Kafka class would also trigger loading the lz4 classes.

ijuma replied:
Yeah, but `KafkaLZ4BlockInputStream` itself should not be initialised until it's actually used. I made this change, used CompressionType.SNAPPY.wrap* methods while LZ4 was not in the classpath and it worked fine. The same exercise did not work for Snappy because `SnappyInputStream` was not in the classpath. Anyway, we can keep it as it is if you prefer as the performance impact is low when compared to other things (8ns instead of 4ns per construction).

xvrl replied:
ok, I've change it to construct the lz4 streams directly, and updated the comments accordingly.

ijuma reviewed (commented):
I did a first pass and left a few comments.

ijuma commented on checkstyle/import-control.xml:
```diff
@@ -50,6 +50,9 @@
     <allow pkg="org.apache.kafka.test" />
 
+    <allow pkg="org.openjdk.jmh.annotations" />
+    <allow pkg="org.openjdk.jmh.infra" />
```
Why do we need this?

xvrl replied:
the benchmark is in the `org.apache.kafka.common.record` package. Now that we don't rely on package protected classes anymore, we can move it to `org.apache.kafka.jmh` and remove this.

ijuma replied:
Sounds good.

xvrl committed:
remove unnecessary synchronized on reset/mark

xvrl commented:
retest this please

xvrl commented:
retest this please

xvrl commented:
Retest this please

xvrl committed:
avoid exception when reaching end of batch

xvrl commented:
retest this please

xvrl commented:
retest this please

xvrl commented:
@ijuma adressed all the comments I believe, now just waiting on getting the build to pass.
@hachikuji you might be interested in reviewing this part https://github.com/apache/kafka/pull/2967/commits/97e3ce81cea2736232a9527dc6a8cbd29e366ccd

xvrl commented:
retest this please

xvrl commented:
Retest this please

xvrl commented:
retest this please

xvrl commented:
retest this please

xvrl commented:
Retest this please

hachikuji commented:
@xvrl Probably not necessary to get a full green build if the failures are known issues. On the other hand, if you're just trying to shake out some more transient failures so that we can add JIRAs, have at it!

hachikuji commented on clients/src/main/java/org/apache/kafka/common/record/AbstractLegacyRecordBatch.java:
```diff
@@ -264,26 +263,48 @@ static void writeHeader(DataOutputStream out, long offset, int size) throws IOEx
+        private static final int OFFSET_OFFSET = 0;
+        private static final int SIZE_OFFSET = 8;
+        private static final int BUFFER_SIZE = 8 // offset
+                                               + 4; // size
```
nit: this looks a little weird. could we just do
```java
private static final int BUFFER_SIZE = 12; // offset + size
```
Also, the name is a bit vague. Maybe `RECORD_HEADER_SIZE`?

ijuma replied:
Yeah, my PR removes all of these in favour of existing constants: https://github.com/apache/kafka/pull/3164/commits/a7bc3e6e98ad3f983f5619601b216e83142afe42

hachikuji commented on clients/src/main/java/org/apache/kafka/common/record/AbstractLegacyRecordBatch.java:
```diff
@@ -264,26 +263,48 @@ static void writeHeader(DataOutputStream out, long offset, int size) throws IOEx
-                long offset = stream.readLong();
-                int size = stream.readInt();
+                int n = 0;
+                while (n < BUFFER_SIZE) {
```
This block seems to be the same code as below. Perhaps we could move it into a `readFully` function in `Utils` (we have a couple of these for working with `FileChannel` already)?

ijuma replied:
You and I had the exact same thought: https://github.com/apache/kafka/pull/3164/commits/a7bc3e6e98ad3f983f5619601b216e83142afe42

xvrl commented:
@hachikuji I give up 😃  I think I've have each Java/Scala combination succeed at least  once. Also haven't seen any new failures in the last few runs.

xvrl commented:
retest this please

xvrl committed:
remove reflection for LZ4 and add comments

xvrl commented:
@ijuma I have addressed all your comments

ijuma committed:
Simplify DataLogInputStream.nextBatch

ijuma committed:
Minor comment improvement

xvrl committed:
ensure ThreadLocal value doesn't maintain a reference to the ThreadLocal key

Anonymous inner BufferSupplier class references the ThreadLocal instance (key) itself

xvrl commented:
Retest this please

xvrl commented:
Retest this please

ijuma committed:
Minor tweaks in `DefaultRecord.readFrom`

ijuma committed:
Cache decompression buffers in Fetcher instead of thread-locals

This means that this only benefits the consumer for now, which
is the most important case. For the server, we should consider
how this fits with KIP-72.

ijuma committed:
Tweaks to `KafkaLZ4*Stream` classes and `RecordBatchIterationBenchmark

ijuma commented on clients/src/main/java/org/apache/kafka/common/record/KafkaLZ4BlockInputStream.java:
```diff
@@ -65,31 +77,21 @@
-        checksum = XXHashFactory.fastestInstance().hash32();
+    public KafkaLZ4BlockInputStream(ByteBuffer in, BufferSupplier bufferSupplier, boolean ignoreFlagDescriptorChecksum) throws IOException {
         this.ignoreFlagDescriptorChecksum = ignoreFlagDescriptorChecksum;
+        this.in = in.order(ByteOrder.LITTLE_ENDIAN);
```
Is it safe to do this on the received buffer? `DeepRecordsIterator` doesn't seem to duplicate the buffer before calling `CompressionType.wrapForOutput`.

xvrl replied:
`DeepRecordsIterator` technically duplicates the buffer when calling [`wrapperRecord.value()`](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/record/AbstractLegacyRecordBatch.java#L310) and never reuses it, but I agree that could easily be missed or introduce odd bugs if that code were to be rewritten. I can add a `duplicate()` here just for safety.

ijuma replied:
Sounds good.

ijuma commented on clients/src/main/java/org/apache/kafka/common/record/KafkaLZ4BlockInputStream.java:
```diff
@@ -145,46 +154,72 @@ private void readHeader() throws IOException {
+            throw new IOException(PREMATURE_EOS);
+        }
+
+        int blockSize = in.getInt();
```
I'm not totally sure why we were previously using a `readUnsigned` method and now we're not. For many of the other cases where we do this, it's easy to verify that the values we care about work either way. The ones that weren't totally clear to me were this and the block checksum cases. Can you please elaborate why the change is safe in those two cases?

xvrl replied:
from reading the code, `ByteUtils.readUnsignedIntLE` is a misnomer and should probably be just be called `readIntLE`. Unlike `readUnsignedInt`, which returns a long and applies the proper bit mask, `readUnsignedIntLE` returns an integer and doesn't do anything other than shifting the bytes.

ijuma commented on clients/src/main/java/org/apache/kafka/common/record/KafkaLZ4BlockInputStream.java:
```diff
@@ -145,46 +154,72 @@ private void readHeader() throws IOException {
+
+        int blockSize = in.getInt();
+        boolean compressed = (blockSize & LZ4_FRAME_INCOMPRESSIBLE_MASK) == 0;
+        blockSize &= ~LZ4_FRAME_INCOMPRESSIBLE_MASK;
```
Hmm, so this bug meant that the `blockSize > maxBlockSize` check was not working, right? Also, did we just get lucky that the `blockSize == 0` check worked?

xvrl replied:
Indeed, the size check wasn't working, but at the end of the day, we would just thrown a different exception in case the block size exceeded the max.

Regarding the zero size block, we were just lucky that we never set the `LZ4_FRAME_INCOMPRESSIBLE_MASK` flag in `KafkaLZ4BlockOutputStream.writeEndMark`. I checked the spec and it didn't seem to specify whether the flag should be set or not in that case, so better be safe.

ijuma commented on clients/src/test/java/org/apache/kafka/common/record/KafkaLZ4Test.java:
```diff
@@ -42,37 +50,135 @@
+        final ByteBuffer buffer = ByteBuffer.wrap(compressed).order(ByteOrder.LITTLE_ENDIAN);
+
+        int blockSize = buffer.getInt(7);
+        blockSize = (blockSize & LZ4_FRAME_INCOMPRESSIBLE_MASK) | (1 << 24 & ~LZ4_FRAME_INCOMPRESSIBLE_MASK);
```
Nit: is there a reason why we are modifying the existing block size instead of just setting the bad block size directly?

xvrl replied:
yes, to make sure it works in both cases when the incompressible flag is set and when it is not.

ijuma committed:
Trivial style tweaks to KafkaLZ4Test

ijuma committed:
Provide a `NO_CACHING` BufferSupplier

ijuma reviewed (commented):
Sorry for the delay, I finally had a chance to complete my review. I left 4 comments questions in this PR and submitted a separate PR with an alternative to the thread local buffers:

https://github.com/apache/kafka/pull/3164

I would normally do the PR with your branch as the target, but because I rebased my branch on trunk (and fixed the merge conflicts), this would not have the desired effect. Please take a look at the other PR and we can decide what we should pull into this PR.

ijuma committed:
Fix checkstyle errors

ijuma committed:
Add parameter names to `MemoryRecordsTest`

xvrl committed:
preserve endianness on input buffer

hachikuji commented on clients/src/main/java/org/apache/kafka/common/record/CompressionType.java:
```diff
@@ -179,4 +185,17 @@ private static MethodHandle findConstructor(String className, MethodType methodT
+                theBuffer = ByteBuffer.allocate(size);
+            }
+            theBuffer.limit(size);
+            return theBuffer;
```
Should we rewind the buffer in case the position wasn't reset after the last use?

(Also nit: can we just call it `buffer`?)

ijuma replied:
That's one issue. The other is that if there are two `KafkaLZ4BlockInputStream` instances in the same thread: https://github.com/apache/kafka/pull/3164/commits/8403cbb616dd88cd7b06dfe8910aa0d0cded8b6f#diff-ee888ddca9dfb15863eeddcd379f65c6R36

xvrl replied:
@hachikuji technically we don't have to, KafkaLZ4BlockInputStream resets position and limit every time anyway when preparing the buffer for consumption.

@ijuma agree that would be a concern if someone were to consume from two separate consumers in the same thread. Happy to replace the buffer supplier implementation with the one you pointed to.

hachikuji replied:
@xvrl Yeah, figured that was the case, but it makes the code a bit brittle to depend on that.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/record/CompressionType.java:
```diff
@@ -179,4 +185,17 @@ private static MethodHandle findConstructor(String className, MethodType methodT
         }
     }
 
+    static class RecyclingBufferSupplier implements KafkaLZ4BlockInputStream.BufferSupplier {
```
It's a little odd for this to live here given the interface is in `KafkaLZ4BlockInputStream`. I'd suggest either moving it there or pulling the `BufferSupplier` interface out of `KafkaLZ4BlockInputStream`.

ijuma replied:
3 for 3: https://github.com/apache/kafka/pull/3164/commits/8403cbb616dd88cd7b06dfe8910aa0d0cded8b6f#diff-ee888ddca9dfb15863eeddcd379f65c6R1

hachikuji reviewed (commented):
Thanks for the patch. I didn't look deeply at the changes in `KafkaLZ4BlockInputStream`, but I left a few comments in `AbstractLegacyRecordBatch`.

xvrl committed:
minor naming changes

xvrl commented:
retest this please

ijuma commented:
Client system tests build will appear in the following link when it starts: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/903/

ijuma commented:
Clients system tests passed: https://jenkins.confluent.io/job/system-test-kafka-branch-builder-2/304/

ijuma reviewed (approved):
Thanks for the PR, this is a huge performance improvement. :)

ijuma commented:
LGTM, merging to trunk and 0.11.0.
